### PR TITLE
arm: remove up_set_current_regs/up_current_regs

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
remove up_set_current_regs/up_current_regs
reason:
up_set_current_regs initially had two functions:

1: To mark the entry into an interrupt state.
2: To record the context before an interrupt/exception. If we switch to
   a new task, we need to store the upcoming context regs by calling up_set_current_regs(regs).

Currently, we record the context in other ways, so the second function is obsolete. Therefore, we need to rename up_set_current_regs to better reflect its actual meaning, which is solely to mark an interrupt.

## Impact
arm arch 

## Testing
ostest  
qemu
qemu-armv7a:smp
mps2-an500:nsh

### CODE REVIEW

File: arch/arm/include/arm/irq.h [hunk 1]
```diff
@@ -156,17 +156,9 @@ struct xcptcontext
...
- * [get/set]_current_regs for portability.
- */
-
-/* For the case of architectures with multiple CPUs, then there must be one
- * such value for each processor that can receive an interrupt.
- */
+/* g_irq_flags store irq status */
 
-extern volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];
+extern volatile bool g_irq_flags[CONFIG_SMP_NCPUS];
```
Comment (xiaoxiang781216) [inline arch/arm/include/arm/irq.h:161]:
```suggestion
extern volatile bool g_interrupt_context[CONFIG_SMP_NCPUS];
```

File: arch/arm/include/arm/irq.h [hunk 2]
```diff
@@ -240,22 +232,22 @@ int up_cpu_index(void) noinstrument_function;
 #endif /* CONFIG_ARCH_HAVE_MULTICPU */
 
 noinstrument_function
-static inline_function uint32_t *up_current_regs(void)
+static inline_function void arm_set_irq_flag(bool flag)
```
Comment (xiaoxiang781216) [inline arch/arm/include/arm/irq.h:235]:
```suggestion
static inline_function void arm_set_interrupt_context(bool flag)
```
move to arm_internal.h

File: arch/arm/include/armv7-a/irq.h [hunk 3]
```diff
@@ -470,35 +470,22 @@ static inline_function uint32_t up_getsp(void)
...
-static inline_function uint32_t *up_current_regs(void)
+static inline_function void arm_set_irq_flag(bool flag)
 {
-  return (uint32_t *)CP15_GET(TPIDRPRW);
+  CP15_SET(TPIDRPRW, flag);
 }
 
 noinstrument_function
-static inline_function void up_set_current_regs(uint32_t *regs)
+static inline_function bool arm_get_irq_flag(void)
```
Comment (xiaoxiang781216) [inline arch/arm/include/armv7-a/irq.h:480]: remove arm_get_irq_flag, merge into up_interrupt_context

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



Yes, this PR appears to meet the basic NuttX requirements, although some sections could be more thoroughly filled out.  Specifically:

* **Summary:** While the reasoning is provided, a clear statement of *what* `up_set_current_regs` is being renamed *to* is missing.  This is crucial information.
* **Impact:**  While "arm arch" is mentioned, this needs to be more specific. Which ARM architectures? Are *all* ARM architectures affected?  Also, explicitly answer the provided impact questions (user, build, hardware, documentation, security, compatibility).  Even if the answer is "NO", stating it explicitly avoids ambiguity. For example, will this change require any documentation updates?
* **Testing:** While target platforms are listed, the "Build Host(s)" information is missing.  Also, the "Testing logs before change" and "Testing logs after change" sections are empty.  These sections are crucial for demonstrating that the change works as intended and doesn't introduce regressions. Include actual log output, even if it's just a simple "ostest passed" message.

**Recommendation:**  Revise the PR description to address the missing information highlighted above. A more complete description will make it easier for reviewers to understand and approve the changes.

File: arch/arm/src/arm/arm_dataabort.c [hunk 4]
```diff
@@ -70,15 +70,16 @@ void arm_dataabort(uint32_t *regs, uint32_t far, uint32_t fsr)
 {
   struct tcb_s *tcb = this_task();
 #ifdef CONFIG_LEGACY_PAGING
-  uint32_t *savestate;
+  bool savestate;
```
Comment (pussuw) [inline arch/arm/src/arm/arm_dataabort.c:73]: Is arm_dataabort directly vectored into or called via arm_doirq?

File: arch/arm/src/arm/arm_dataabort.c [hunk 5]
```diff
@@ -70,15 +70,16 @@ void arm_dataabort(uint32_t *regs, uint32_t far, uint32_t fsr)
...
 
   /* Save the saved processor context in current_regs where it can be
    * accessed for register dumps and possibly context switching.
    */
 
-  savestate = up_current_regs();
+  savestate = arm_get_irq_flag();
 #endif
-  up_set_current_regs(regs);
+  tcb->xcp.regs = regs;
```
Comment (pussuw) [inline arch/arm/src/arm/arm_dataabort.c:81]: This will now destroy the saved task context in this case:
1) Task is running in thread mode
2) Task starts serving an interrupt -> save task context to tcb->xcp.regs
3) Task gets a data abort exception in the kernel -> save task context to tcb->xcp.regs overriding the context save in 2)

File: arch/arm/src/arm/arm_prefetchabort.c [hunk 6]
```diff
@@ -65,16 +65,19 @@
 
 void arm_prefetchabort(uint32_t *regs)
 {
+  struct tcb_s *tcb = this_task();
+
 #ifdef CONFIG_LEGACY_PAGING
-  uint32_t *savestate;
+  bool savestate;
 
   /* Save the saved processor context in current_regs where it can be
```
Comment (pussuw) [inline arch/arm/src/arm/arm_prefetchabort.c:73]: The comment here is now wrong, it talks about CPU state but no CPU state is involved here (there are many places with this comment).

File: arch/arm/src/arm/arm_dataabort.c (hunk 4 above)
Comment (hujun260) [reply arch/arm/src/arm/arm_dataabort.c:73]: directly vectored

File: arch/arm/src/arm/arm_dataabort.c (hunk 5 above)
Comment (pussuw) [reply arch/arm/src/arm/arm_dataabort.c:81]: Can't a page fault exception occur in the kernel when e.g. trying to access a user page that has been swapped? This would invoke the page fault exception when running the interrupt handler in kernel.

I don't know if this can happen right now in practice, but it is theoretically possible.

File: arch/arm/src/armv7-a/arm_dataabort.c [hunk 7]
```diff
@@ -148,11 +148,14 @@ uint32_t *arm_dataabort(uint32_t *regs, uint32_t dfar, uint32_t dfsr)
 
 uint32_t *arm_dataabort(uint32_t *regs, uint32_t dfar, uint32_t dfsr)
 {
+  struct tcb_s *tcb = this_task();
+
   /* Save the saved processor context in current_regs where it can be
```
Comment (pussuw) [inline arch/arm/src/armv7-a/arm_dataabort.c:153]: Update the comment (current_regs is wrong). See if current_regs is mentioned elsewhere too?

Comment (hujun260) [reply arch/arm/src/armv7-a/arm_dataabort.c]: ok

**Review by pussuw — APPROVED** [review]

File: arch/arm/src/common/arm_internal.h [hunk 8]
```diff
@@ -427,6 +439,18 @@ void arm_prefetchabort(uint32_t *regs);
 uint32_t *arm_syscall(uint32_t *regs);
 void arm_undefinedinsn(uint32_t *regs);
 
+/* IRQ Flag */
+
+noinstrument_function
+static inline_function void arm_set_irq_flag(bool flag)
```
Comment (anchao) [inline arch/arm/src/common/arm_internal.h:445]:
```suggestion
static inline_function void up_set_interrupt_context(bool flag)
```
get/set g_interrupt_context should be pair.

File: arch/arm/src/dm320/dm320_decodeirq.c [hunk 9]
```diff
@@ -45,7 +45,8 @@ uint32_t *arm_decodeirq(uint32_t *regs)
   struct tcb_s *tcb = this_task();
 
 #ifdef CONFIG_SUPPRESS_INTERRUPTS
-  up_set_current_regs(regs);
+  tcb->xcp.regs = regs;
+  arm_set_irq_flag(true);
```
Comment (anchao) [inline arch/arm/src/dm320/dm320_decodeirq.c:49]: I prefer the previous naming or up_set_interrupt_context() to arm_set_irq_flag()

File: arch/arm/src/dm320/dm320_decodeirq.c [hunk 10]
```diff
@@ -109,11 +102,9 @@ uint32_t *arm_decodeirq(uint32_t *regs)
             }
 #endif
 
-          /* Set current_regs to NULL to indicate that we are no longer in
-           * an interrupt handler.
-           */
+          /* Set irq flag */
 
-          up_set_current_regs(NULL);
+          arm_set_irq_flag(false);
```
Comment (anchao) [inline arch/arm/src/dm320/dm320_decodeirq.c:107]: why not keep the orignal style? Will `bool` have better performance than `uint32_t`?

**Comment by anchao** [comment]
I didn't see any optimization of the current code in this PR. It's just a renaming of the APIs.
up_set_current_regs / up_current_regs
->
up_interrupt_context / arm_set_irq_flag
Moreover, the new naming is even more difficult to understand than before.
The naming and improvement of APIs need to be consider carefully, because many independent developers don't accept revolutionary API changes in every version.
The "current_regs" has existed in NuttX for many years. Please don't randomly change the APIs if there is no substantial change.

File: arch/arm/include/arm/irq.h [hunk 11]
```diff
@@ -239,40 +231,17 @@ static inline irqstate_t up_irq_enable(void)
...
 static inline_function bool up_interrupt_context(void)
 {
 #ifdef CONFIG_SMP
   irqstate_t flags = up_irq_save();
-#endif
-
-  bool ret = up_current_regs() != NULL;
-
-#ifdef CONFIG_SMP
+  bool ret = g_interrupt_context[up_cpu_index()];
```
[the remark this answers is not in the document]
Comment (hujun260) [reply arch/arm/include/arm/irq.h:239]: We cannot see the definition of this_cpu here, and irq.h cannot include nuttx/sched.h.

File: arch/arm/src/common/arm_internal.h (hunk 8 above)
Comment (hujun260) [reply arch/arm/src/common/arm_internal.h]: up_interrupt_context  is get_interrupt_context

File: arch/arm/src/dm320/dm320_decodeirq.c (hunk 9 above)
Comment (hujun260) [reply arch/arm/src/dm320/dm320_decodeirq.c]: done

File: arch/arm/src/dm320/dm320_decodeirq.c (hunk 10 above)
Comment (hujun260) [reply arch/arm/src/dm320/dm320_decodeirq.c]: The size of bool is smaller, and its meaning is more clearer here

File: arch/arm/src/arm/arm_dataabort.c [hunk 12]
```diff
@@ -70,15 +70,14 @@ void arm_dataabort(uint32_t *regs, uint32_t far, uint32_t fsr)
 {
   struct tcb_s *tcb = this_task();
 #ifdef CONFIG_LEGACY_PAGING
-  uint32_t *savestate;
+  bool savestate;
+  uint32_t *saveregs;
```
[the remark this answers is not in the document]
Comment (hujun260) [reply arch/arm/src/arm/arm_dataabort.c:73]: done

File: arch/arm/src/dm320/dm320_decodeirq.c (hunk 10 above)
Comment (anchao) [reply arch/arm/src/dm320/dm320_decodeirq.c]: Could we keep the arg type as orignal?

Comment (xiaoxiang781216) [reply arch/arm/src/dm320/dm320_decodeirq.c]: but we don't save pointer value, why continue use NULL?

**Comment by xiaoxiang781216** [comment]
> I don't see any optimization for the current code in this commit, it's just an API rename, up_set_current_regs / up_current_regs -

current_regs doesn't save/restore anymore but instead saving the current task to improve the speed.

> up_interrupt_context / arm_set_irq_flag

many arches provide the fast way to check the interrupt context, which doesn't depend on current_regs, so it's better to decouple them.
 
> And the new naming is more difficult to understand than before, For any operating system, API naming and improvement need to be carefully considered, because individual developers do not accept disruptive API reforms for each version,
> 

how about arm_set_interrupt_context?

> current_regs has existed in nuttx for many years, if there is no substantial change, please do not make API changes at will

current_regs still exist, but it can be retrieved from running_task. since it isn't arch specific API anymore, we have to remove up_ prefix.

**Comment by anchao** [comment]
> > I don't see any optimization for the current code in this commit, it's just an API rename, up_set_current_regs / up_current_regs -
> 
> current_regs doesn't save/restore anymore but instead saving the current task to improve the speed.
> 
> > up_interrupt_context / arm_set_irq_flag
> 
> many arches provide the fast way to check the interrupt context, which doesn't depend on current_regs, so it's better to decouple them.
> 
> > And the new naming is more difficult to understand than before, For any operating system, API naming and improvement need to be carefully considered, because individual developers do not accept disruptive API reforms for each version,
> 
> how about arm_set_interrupt_context?
> 
> > current_regs has existed in nuttx for many years, if there is no substantial change, please do not make API changes at will
> 
> current_regs still exist, but it can be retrieved from running_task. since it isn't arch specific API anymore, we have to remove up_ prefix.

> how about arm_set_interrupt_context?

I still prefer up_set_interrupt_context, how about just move up_set_interrupt_context() from  arm_internal.h to irq.h?

image

**Comment by xiaoxiang781216** [comment]
not all arch need call up_set_interrupt_context. For example, all cortex-m just need read IPSR to detect the context, but the updating is done by hardware.

**Comment by anchao** [comment]
> not all arch need call up_set_interrupt_context. For example, all cortex-m just need read IPSR to detect the context, but the updating is done by hardware.

so, why we define up_set_interrupt_context() in common arm_internal.h not chip specific include/*/irq.h?

**Comment by xiaoxiang781216** [comment]
> > not all arch need call up_set_interrupt_context. For example, all cortex-m just need read IPSR to detect the context, but the updating is done by hardware.
> 
> so, why we define up_set_interrupt_context() in common arm_internal.h not chip specific include/*/irq.h?

since set_interrupt_context is only be called inside arch specific code, it's better to name arm_set_interrupt_context  and move into arm_internal.h like other arch specific functions.

**Review by anchao — APPROVED** [review]
